src/runtime/pipeline/pipeline_struct.h

```diff
@@ -42,6 +52,75 @@ using GlobalOutputPair = std::pair<int, int>;
  * The first 'int' is the module index, and the second 'int' is the module output index.
  */
 using ModuleOutputPair = std::pair<int, int>;
+/*!
+ *\brief The pair includes the module index and the module input index.
+ * The first 'int' is the module index, and the second 'int' is the module input index.
+ */
+using ModuleInputPair = std::pair<int, int>;
+/*!\brief The data notification structure.*/
+class DataNotify {
+ private:
+  /*!\brief The 'contitional variable' is used to wait for notification.*/
+  std::condition_variable condition_;
+  /*!\brief The mutex is used to protect the 'conditional variable'.*/
+  std::mutex mutex_;
+  /*!\brief Whether a data is ready or not.*/
+  volatile bool data_ready_ = false;
+  /*!\brief Whether the thread should exit or not.*/
+  volatile bool exit_state_ = false;
```

Comment:
seems like not need to use atomic, because 'data_ready_' already protected by 'mutex_'  and for exit_state, there is only a single 'writing' in ExitNotify then no race condition can happen.  although atomic is very light weight but it still involved additional perf cost, if not necessary we may need to avoid use it here.

Reply:
> seems like not need to use atomic, because 'data_ready_' already protected by 'mutex_' and for exit_state, there is only a single 'writing' in ExitNotify then no race condition can happen

Then why do you need `volatile`? I'm not really familiar with `volatile` except that it is generally considered as a bad practice.

Reply:
I agree with that volatile is bad practice for 'atomic' guarantee, but in our case we only need to guarantee the data 'visibility' under multiple thread mode ,  for such requirement volatile should be ok.

about 'atomic' guarantee part, for vairable 'data_ready_' the existing 'mutex_' and conditional variable will help for the atomic guarantee,   for variable exit_state_, because each notification only have one parent and one child, under such single consumer and single producer scenario the both reading and writing of 'exit_state' are native atomic then no need  additional protection.

Reply:
I don't know, but all of SO answers I looked at say volatile is not needed for threading purpose. For example https://stackoverflow.com/questions/4557979/when-to-use-volatile-with-multi-threading 

Any thought?

Reply:
volatile has nothing to do with multi-threading or memory fence. it's only used in modern c++ to enforce reading from memory and disallow compilers to put the specific data from the register. Note that no memory fence is required to be introduced by volatile, so it doesn't officially guarantee anything about visibility. The rule of thumb is to never use volatile unless for usecases like DMA

Reply:
@huajsj what do you think about using something like `std::promise` instead? while I think the use of volatile here is probably okay in a consumer-producer relation, i'm wondering if we can use a more established primitive to enable more folks in the community to work on the code.

Reply:
@masahi ,@junrushao1994 , thanks for the follow up, yes I agree with that std::atomic should be the correct solution to guarantee MT instead of using 'volatile' which may still need a memory barrier solution.

the latest change removed volatile for data_ready_ because mutex already did the  acquire/release memory barrier, and change exit_state_ into a std::atomic<bool>, please check.

Reply:
@masahi ,@junrushao1994, @areusch  , thanks for the follow up,  I agree with that the promise is a better primitive and memory fence also be another issue we need to correct handle, after review code, I just realized there is still a  memory barrier need to put after set 'exit_state = true' and before 'notify()' to guarantee the code 'ordering' , for such issue I may either use std::atomic  or use 'promise + release barrier' and seems like std::atomic should be the better solution with one logic.

in the latest change I removed volatile for data_ready_ because mutex already did the  acquire/release memory barrier, and change exit_state_ into a std::atomic<bool>, please check.